Worth trying to merge? Major additions/changes related to OpenAPI spec generation

jlevers opened the pull request:
Hey @HelgeSverre,

Thanks for all your hard work on this package.

A few months ago I started using it as the basis for generating [jlevers/selling-partner-api](https://github.com/jlevers/selling-partner-api), and ended up making major modifications to, mostly to make it cover more of the OpenAPI spec, which is what I needed it for. I believe this was before you'd added support for generating request body/response body DTOs, or nested DTOs.

I see that since then you've added support for DTO generation, but I think my fork has support for a few things that the upstream repo doesn't: `allOf` schemas, more customizable namespacing via the `Config` class, the ability to configure and run the generator entirely from another PHP file in addition to the command line tool, etc.

Since you've made a lot of changes since I started working on my fork and I've never synced with your changes, obviously merging the two is going to be a major task. I guess I'm just trying to figure out a merge is something you have any interest in before I start doing the work to make it happen. No hard feelings at all if you don't want to merge, I've obviously changed a ton of your original code.

Cheers!

jlevers committed:
Make it possible to load config options from a JSON file

* If CodeGenerator is not passed a Config instance, it will look for a
  file named "generator-config.json" in the current working directory
  and use that as the config file.
* Config::load() loads a JSON config file into a Config instance. If its
  $path argument is null, it will look for a file named "generator-
  config.json" in the current working directory.

jlevers committed:
Create NullGenerator for skipping generation of an asset category

The NullGenerator class doesn't output anything, so if (for instance)
you don't want to generate connectors, you can pass a NullGenerator
instance as the connectorGenerator argument of the Generator class.

jlevers committed:
Remove unused AppServiceProvider

jlevers committed:
Write files to disk from GeneratedCode

Most of the functionality in GenerateSdk that writes generated code to
disk has been moved to GeneratedCode. This allows us to write code to
disk when generating code programmatically, rather than from the command
line. All command-line functionality remains the same, and stayed in
GenerateSdk.

jlevers committed:
Add $overrides param to Config::load() for overriding file config values

jlevers committed:
Use `static` instead of `self` in Parser::build() methods

This makes it possible to extend existing parser classes, and override
a subset of their methods.

jlevers committed:
Pass null to namespace suffix fields to disable them entirely

Previously, there was no way to avoid having namespace suffixes - if a
null value was passed, the default value was used. This commit makes it
so that passing an explicit null value will disable the resource and
request namespace suffixes entirely.

jlevers committed:
Create FileHandler contract for managing output file structure

The CodeGenerator and GeneratedCode classes now take a FileHandler
implementation as a constructor argument. This allows for user-defined
file structure and naming conventions for the generated code.

jlevers committed:
Add baseResourceNamespace config option

jlevers committed:
Allow config file namespace and connectorName to be overridden

jlevers committed:
Basic DTO generator and OpenAPI schema type parser

* Add FileHandler::dtoPath() for specifying the DTO output path
* Implement DtoGenerator::generate()
* Parse schema types from OpenAPI specification
* Create Data\Generator\Schema class for storing parsed schema information

jlevers committed:
Add infrastructure for generating response classes

* Create stubbed ResponseGenerator class
* Add FileHandler::responsePath() method to get the response output path
* Add responseGenerator param to CodeGenerator constructor
* Add resourceClasses param to GeneratedCode constructor, and save those classes when calling dumpFiles()

jlevers committed:
Avoid creating DTO files for scalar/array properties

jlevers committed:
Implement basic response generator

jlevers committed:
Refactor existing generators to use Generator::makeClass()

jlevers committed:
Add DTO and Response generator output to CLI command

jlevers committed:
Make responses and DTOs deserializable from JSON

* Create Deserializable interface and Deserializes trait
* Make basic typing system for DTO attributes

jlevers committed:
Deserialization working!!

Also created BaseDto and BaseResponse classes, which are extended by all
generated Dto and Response classes, respectively. This does two things:
* Avoid implementing the same interfaces and using the same traits in
  generated classes
* Allow us to override the Deserializes trait's static $complexArrayTypes
  in generated Dto and Response classes. You can't override a trait's
  attribute in the class that uses it, but you *can* override it in a
  child of the class that uses it

jlevers committed:
Non-recursive inline schema parsing

By passing parent schema references to child schemas, we can implement
naming and typing for inline child schemas. These changes only cover
single-level child schemas in responses and other schemas.

These changes also include support for scalar array types.

jlevers committed:
Add strict typing to all files

jlevers committed:
Misc small bugfixes

* Make sure __construct method is at the beginning of request classes
* Cast file_put_contents return value to bool in GeneratedCode::dumpToFile()
* Call toString() on result of NameHelper::normalize()

jlevers committed:
Support empty responses

jlevers committed:
Ensure required parameters always come first in resource/request methods

jlevers committed:
Move complex array typing out of Deserializes into HasComplexArrayTypes

We need to handle complex array types in both deserialization and
array-ification of DTOs. This moves the $complexArrayTypes property out
of Deserializes and into a new trait HasComplexArrayTypes, because
otherwise that property would be duplicated in both Deserializes and the
new trait we're going to create to handle turning DTOs into arrays.

jlevers committed:
Make HasArrayableAttributes trait for turning DTOs into arrays

jlevers committed:
Parse OpenAPI request bodies and generate code for them

Request bodies are now assigned to DTOs, which can be passed via
resource methods and request constructors.

jlevers committed:
Compare required fields to original parameter names, not sanitized names

jlevers committed:
Ensure optional parameters always come after required parameters in DTOs

jlevers committed:
Remove unneeded support for defining key types on complex array types

jlevers committed:
Improve Parameter -> Schema compatibility

jlevers committed:
Parse additionalProperties from OpenAPI schema

jlevers committed:
Rename addParameterAsPromotedProperty to addParameterToConstructor

Also, adapt the method to be more flexible to handling different sorts
of parameters and parameter types (including non-promoted parameters).

jlevers committed:
Include additionalProperties in the generated DTOs

jlevers committed:
Functional name normalization for acryonyms

jlevers committed:
Switch SimpleType::NUMBER to SimpleType::FLOAT

jlevers committed:
Improve complex array type parser detection

jlevers committed:
Improve requirement detection by passing parent key to child schemas

jlevers committed:
Return EmptyResponse on 204

jlevers committed:
Only use normalized schema names in OpenApiParser

Instead of converting to safe variable/class names in the parser, we can
exclusively use normalized names, and then convert to variable/class
names in the generators. This makes it easier to reason about what's
going on in the parser.

jlevers committed:
Improve schema/parameter docstrings

jlevers committed:
Parse null OpenAPI schema requirements properly

jlevers committed:
Only set schema title in parser if schema doesn't have a title yet

jlevers committed:
Fix request constructor parameter types and complex array type classes

jlevers committed:
Generalize MethodGeneratorHelper method for adding parameters to methods

Renamed MethodGeneratorHelper::addParameterToConstructor() to
MethodGeneratorHelper::addParameterToMethod() and made promote default
to false to make the method suitable for adding parameters to any
method.

jlevers committed:
Fix resource method param types by using addParameterToMethod

jlevers committed:
Use MethodGeneratorHelper::addParameterToMethod in ResponseGenerator

jlevers committed:
Handle array responses

jlevers committed:
Create sub-namespace generator methods in Config

This gets rid of a lot of repeated code in the generator classes. There
are now dtoNamespace, requestNamespace, responseNamespace, and resourceNamespace
methods on the Config class.

jlevers committed:
Replace SimpleType::tryFrom type checks with Utils::isBuiltinType

jlevers committed:
Use SimpleType::tryFrom to check types in Util::isBuiltinType

jlevers committed:
Make sure to import any response types that are used as DTO properties

jlevers committed:
Separate additionalProperties parsing into separate method

jlevers committed:
Add support for allOf schema types in OpenApiParser

jlevers committed:
Find response schema types before parsing schemas

jlevers committed:
Move OpenAPI preprocessing to OpenApiNormalizer, inline schemas working

jlevers committed:
Improve additionalProperties support

* Collapse schemas that only contain additionalProperties into typed arrays
* Don't generate DTOs for additionalProperties-only schemas
* Don't add additionalProperties to every DTO by default

jlevers committed:
Add 'parent' to reserved keyword list for normalization

jlevers committed:
Use EmptyResponse for 204/empty responses

jlevers committed:
Make sure to resolve all references during/after schema normalization

jlevers committed:
$complexArrayTypes generating properly in ResponseGenerator

jlevers committed:
Store original parameter/schema names from spec for serialization

jlevers committed:
Fix nullable array docstring generation, don't change list type names

jlevers committed:
Make sure to use EmptyResponse for 204 responses

jlevers committed:
Handle response type aliases gracefully in RequestGenerator

jlevers committed:
Make HasArrayableAttributes::$attributeMap protected instead of public

jlevers committed:
Use attributeMap when deserializing to ensure data keys match DTO keys

jlevers committed:
Namespace response classes correctly in response complexAttributeTypes

jlevers committed:
Use DateTime instead of string whenever possible (RFC3339)

jlevers committed:
Update package name in composer.json

HelgeSverre commented:
Liking a lot of what I see so far!

Have only gotten to skim part of the changes on a phone screen though, so i'll need to take a closer look after the easter holiday.

I"ll get back to you.

jlevers committed:
Simplify Config parameters to consolidate suffixes into a single param

jlevers committed:
Remove all dependencies on this library from the generated code

* Remove BaseDto, BaseResponse, EmptyResponse
* Create BaseDtoGenerator, BaseResourceGenerator, BaseRequestGenerator, BaseResponseGenerator
* Create SupportingFilesGenerator for copying files directly from this library

jlevers committed:
Bump version to 1.1.0

jlevers committed:
Make all generators fully compliant with upstream Generator interface

jlevers committed:
Update to be Laravel 11 compatible

jlevers committed:
Bump version to 2.0.0

jlevers committed:
Add contentType to Schema

dpash committed:
Use DateTimeInterface for parameters

We should use `\DateTimeInterface` for parameters so that we can accept
`\DateTime` or `\DateTimeImmutable`.

Additionally, use the root namespace on phpdoc comments so that tools
like Psalm don't think there's a class in the current namespace.

dpash committed:
Include the body content type in Schema

We only want to add the HasJsonBody trait to requests expecting an
`application/json` body, but this information is not currently available

This commit adds the data into the schema object so we can retrieve it
when generating request objects.

dpash committed:
Include warning files are autogenerated

I've seen a number of PRs that modify generated files. To prevent people
wasting their time, this adds a comment at the top of the generated file
that says that the file has been autogenerated

dpash committed:
Fix caching of schemas when normalizing

We had issues where the wrong type of arrays was being used, due to them
using the property name as the cache key. If there were two arrays called
`item` the second property would use the type of the first.

To fix this, we now cache arrays using `array-<Type>` as the cache key.

Additionally, we were caching items even if they had different
properties or description.

The caching has been extracted into it's own class to keep the logic
separate.

jlevers committed:
Fix bad placeholder character in deserialization type match statement

jlevers committed:
Merge pull request #3 from dpash/include-body-content-type

Include the body content type in Schema

jlevers committed:
Handle base file paths separately in BasicFileHandler via baseOutputPath

jlevers committed:
Access HasArrayableAttributes::$attributeMap statically

jlevers committed:
Replace Literal with sprintf where possible

jlevers committed:
Merge pull request #2 from dpash/use-datetimeinterface

Use DateTimeInterface for parameters

jlevers committed:
Merge pull request #5 from dpash/warn-autogenerated

Include warning files are autogenerated

jlevers committed:
Merge pull request #6 from dpash/fix-schema-caching

Fix caching of schemas when normalizing

jlevers committed:
Misc small fixes

* Remove $unknownKeys from Deserializes trait
* Allow null $namespaceSuffixes in Generator::makeClass()

jlevers committed:
Check for circular references when parsing schemas

jlevers committed:
Create workflow file to automatically validage packages and lint

jlevers committed:
Formatting

jlevers committed:
Remove null values from serialized data

jlevers committed:
Remove composer.lock, remove version from composer.json

jlevers committed:
Merge pull request #8 from highsidelabs/check-for-infinite-ref-loops

Check for circular references when parsing schemas

jlevers committed:
Remove composer.lock

jlevers committed:
Check description when checking for circular references

jlevers committed:
Allow nameless array schemas during parsing

jlevers committed:
Support form, multipart, xml, string, and stream body content types

jlevers committed:
Allow custom datetime formats via Config

jlevers committed:
Handle naming when there are multiple inline schemas with the same name

jlevers committed:
Bump minimum PHP version to 8.2

jlevers committed:
Handle serialization of request bodies that consist of an array of DTOs

jlevers committed:
Switch from ignoredQueryParams/ignoredBodyParams to ignoredParams

jlevers committed:
Add support for header parameters

jlevers committed:
Reformat complex array types to avoid unnecessary nested arrays

jlevers committed:
Fix codegen for complex array responses

jlevers committed:
Only generate defaultHeaders method if there are non-ignored header arguments

jlevers committed:
Check if $other is an array before comparing schema items in Schema::equalsOpenApiSchema()

jlevers committed:
Make ignored header checks case-insensitive

jlevers committed:
Handle 204 responses in OpenApiNormalizer

jlevers committed:
Remove readonly attribute from DTO properties

jlevers committed:
Add constructor to EmptyResponse class to make it deserializable

danielhe4rt commented:
Hey! I spent some time trying to use your fork but I didn't found how... I'm using with PHP 8.4 but seems that some things changed based on the original docs.

Could you write some lines in your fork on how to use it? Thanks!

jlevers committed:
Make it possible to explicitly override supporting files

jlevers commented:
Hey @danielhe4rt – cool that you're trying to use the fork! I didn't write any docs, since I didn't really expect anyone else to use it. There are a *lot* of changes (as you can see), and it will likely be a while before I get around to writing my own docs, but I will throw it on the todo list.

You should be able to use the code in the [`src/Generators`](https://github.com/jlevers/selling-partner-api/blob/main/src/Generator) directory of `jlevers/selling-partner-api` to figure out how it works, though – the `Generator.php` file is where everything starts. Hope that helps!

bob2021 committed:
Wrap body in MultipartValue if content type is multipart/form-data.

jlevers committed:
Fix cyclical schema detection in OpenApiParser

jlevers committed:
Merge pull request #10 from bob2021/fix-multipart

Wrap body in MultipartValue if content type is multipart/form-data

jlevers committed:
Update to Saloon v4
